### DISCUSSION TITLE
Welcome to Bear Discussions!

### DISCUSSION THREAD

**Discussion by rizsotto** [rizsotto/Bear#339] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Peter-Korobeynikov** [comment]
**_Hello!_**
I have two questions:
1) - Not even a question. For successful compilation with nlohmann/json v3. 9.1 in the file
~/my_path/Bear/source/citnames/source/Output.cc - I had to add the third argument is "" (context)
to call the function nlohmann:: json:: out_of_range:: create (around line 192):
 'throw nlohmann::json::out_of_range::create(403, "key 'command' or 'arguments' not found", "");'

2) - I was never able to get a successful build with spdlog v1.8.2 & fmt v7.1.3 and unit|func -tests:
cmake DCMAKE_INSTALL_PREFIX=$MY_INSTALL_DIR ..
[cmake_out.txt](https://github.com/rizsotto/Bear/files/6148514/cmake_out.txt)

cmake --build . --parallel $(nproc)
...
/usr/bin/ld: CMakeFiles/sys_a.dir/source/Process.cc.o: in function 'void spdlog::logger::log_<fmt::v7::basic_string_view<char>, int, (anonymous namespace)::Arguments> (spdlog::source_loc, spdlog::level::level_enum, fmt::v7::basic_string_view<char> const&, int&&, (anonymous namespace)::Arguments&&) [clone .constprop.0]':
Process.cc:(.text+0x346): undefined reference to 'spdlog::details::backtracer::enabled() const'
... etc ...
error: ld execution ended with return code 1
[build_out.txt](https://github.com/rizsotto/Bear/files/6148517/build_out.txt)

pkg-config spdlog --cflags 
	>> -DSPDLOG_COMPILED_LIB -I/home/master/.local/include
	
(At the same time, cmake with the parameters -DENABLE_UNIT_TESTS=OFF -DENABLE_FUNC_TESTS=OFF gives success)

**_Thanks!_**

**Reply by rizsotto** [reply]
Hey Peter,

For the first item, if you need to add a third parameter to the `nlohmann::json::out_of_range::create` method, you probably was not compiled against v3.9.1 of nlohmann::json. Could you check if your system has an older version installed? That might have been picked up by the compiler. (The method has two overloaded declaration, both takes 2 parameters in v3.9.1)

For the second problem it would be better to have a more verbose log. `VERBOSE=1 cmake --build .` (Don't use `--parallel` it makes the logs harder to read.)

But what your questions are? :)

It would be also good to know what OS you are compiling for? What is the CMake version you use? How the dependencies were installed on your system?

**Comment by Peter-Korobeynikov** [comment]
**_Hi, Laszlo!_**

While on question 1... 
I'm worried about my weak level, so I'll say it easier :)

OS =Debian GNU/Linux 10 (buster)
cat /proc/version
>> Linux version 4.19.0-14-amd64 (debian-kernel@lists.debian.org) (gcc version 8.3.0 (Debian 8.3.0-6)) #1 SMP
Debian 4.19.171-2 (2021-01-30)

GCC is built locally (~/. local)
gcc --version
>> gcc (GCC) 9.3.0

In the profile:
<pre>
        export MY_INSTALL_DIR="$HOME/.local"
        ...
	if [ -d $HOME/.local/include ] ; then
	    export C_INCLUDE_PATH="$C_INCLUDE_PATH:$HOME/.local/include"
	    export CPLUS_INCLUDE_PATH="$CPLUS_INCLUDE_PATH:$HOME/.local/include"
	    export CPATH="$C_INCLUDE_PATH"
	fi
</pre>

Installing json:
cd ~/Install/Tools
git clone https://github.com/nlohmann/json
cd json && mkdir -p build && cd build
cmake -DCMAKE_INSTALL_PREFIX=$MY_INSTALL_DIR ..
make -j$(nproc) && make install

File ~/. local/include/nlohmann/json.hpp attached:
[json.zip](https://github.com/rizsotto/Bear/files/6161807/json.zip)


See line #2957
<pre>
class out_of_range : public exception
{
   public:
   template<typename BasicJsonType>
   static out_of_range create(int id_, const std::string& what_arg, const BasicJsonType& context)
   {
      std::string w = exception::name("out_of_range", id_) + exception::diagnostics(context) + what_arg;
      return out_of_range(id_, w.c_str());
   }
   private:
   JSON_HEDLEY_NON_NULL(3)
   out_of_range(int id_, const char* what_arg) : exception(id_, what_arg) {}
};
</pre>
My compiler gets here (I tried changing the json.hpp code). And this is version 3.9.1 (!)
I only see a private constructor with two arguments.
So I can't immediately agree that v3.9.1 has an overloaded function with two arguments.

Thank you for your work!
=Peter &#128522;

**Reply by rizsotto** [reply]
Hi Peter,

thanks to come back. I am happy to see you did not give up on this. ;)

Maybe you know it already, but Bear's build is written in a way that it can compile its dependencies from source, if those are missing. (One exception is SQLite.) That is a simpler way to build it and might eliminate the potential errors. So, if you remove `fmt`, `spdlog`, `grpc` and `nlohmann-json` it still builds it. (You don't need to manually build those beforehand. Version match is guaranteed.) I would recommend to give a try to build it this way.

The CI job of the project is using Ubuntu 20.4, which I guess not far from the Debian system you are using. So, you can look at [how the CI runs the build](https://github.com/rizsotto/Bear/blob/master/.github/workflows/build_on_push.yml). It also just installs the build tools, but not the dependencies.

**Comment by Peter-Korobeynikov** [comment]
Yes, I haven't given up on my idea.
I'm doing a port autotools->cmake of one project for future development.
Yes, I know that bear build the necessary tools.
Similarly, I tried to solve my second problem with the bear build (in which ld can't find a link for spdlog functions). But even in the case of automatic installation Bear by cmake, this error did not disappear.

As for json, I think it will be the same result. I haven't tried it though. I may try to do this again for json. But I think that the result will be the same as with manual build. It will be interesting to try it. But I think that the third argument is still there in nlohmann:: json::out_of_range:: create. Anyway :)
(!) By the way, this is a new situation. A month ago, Bear compiled without problems. But that was before the update from Debian-9 to Debian-10. (Perhaps something has changed in the dependencies over the past month)

Despite these build inaccuracies, bear still built the compilation database. And the json2cmake project created a file CMakeLists.txt, which read and built successfully by CLion.
So I have already jump through this problems.
Thanks.

**Comment by jgcodes2020** [comment]
Will this tool ever get Windows support? I use Gradle and my attempts to extract compiler information in the build script doesn't work too well.

**Reply by rizsotto** [reply]
Hey @jgcodes2020 , good question. I won't be able to do it myself. So, it's up to the community. PRs are welcome!

**Comment by lol5230** [comment]
I saw that gRPC was a dependency. Can bear run without a connection to the internet? I'm looking to possibly run this on a server of sorts that would be on an intranet not connected to the internet.

**Reply by rizsotto** [reply]
Hey @lol5230 , running Bear without Internet connection is no problem. The gRPC is used to communicate between processes within the host. It binds to `127.0.0.1` address only.
